internal/pkg/service/common/etcdop/op/atomic.go

```diff
@@ -31,26 +26,51 @@ import (
 // - The DoWithoutRetry method returns false.
 //
 // Retries on network errors are always performed.
+//
+// The atomic operation can be read from the context during generation to be extended, see AtomicFromCtx function.
+// This can lead to several subsequent READ phases, but there is always only one WRITE phase, after all READ phases.
 type AtomicOp[R any] struct {
 	*AtomicOpCore
-	result         *R
-	processors     processors[R]
-	checkPrefixKey bool // checkPrefixKey - see SkipPrefixKeysCheck method documentation
-}
-
-// mutex abstracts concurrency.Mutex and etcdop.Mutex types.
-type mutex interface {
-	Key() string
-	IsOwner() etcd.Cmp
+	result     *R
+	processors processors[R]
 }
 
 type AtomicOpInterface interface {
+	// ReadPhaseOps returns all op factories for READ phase,
+	// is used in joining two atomic operations, see AddFrom method.
 	ReadPhaseOps() []HighLevelFactory
+	// WritePhaseOps returns all op factories for WRITE phase,
+	// is used in joining two atomic operations, see AddFrom method.
 	WritePhaseOps() []HighLevelFactory
+	// Core returns AtomicOpCore - This does not contain result type specific methods.
+	Core() *AtomicOpCore
 }
 
+// Mutex abstracts concurrency.Mutex and etcdop.Mutex types.
+type Mutex interface {
+	Key() string
+	IsOwner() etcd.Cmp
+}
+
+// Atomic creates empty atomic operation with result type R.
 func Atomic[R any](client etcd.KV, result *R) *AtomicOp[R] {
-	return &AtomicOp[R]{AtomicOpCore: &AtomicOpCore{client: client}, result: result, checkPrefixKey: true}
+	v := &AtomicOp[R]{AtomicOpCore: newAtomicCore(client), result: result}
+	v.setProcessorFactory(func() func(ctx context.Context, r *TxnResult[NoResult]) {
+		if v.processors.len() == 0 {
+			return nil
+		}
+
+		return func(ctx context.Context, r *TxnResult[NoResult]) {
+			if r.Succeeded() || r.Err() != nil {
```

Comment:
It is some weird typo, that condition does not have to be there at all, it should only add the `v.result`, and use new typed result. Fixed.

(since OR is in that condition, it is practically always performed)